### PR TITLE
ImageMatching new mode `a+ab`

### PR DESCRIPTION
**Pull request by gregoire-dl**
## Features list
- [x] rename image matching modes
- [x] add image matching mode `a/a+a/b`

### CODE REVIEW

File: src/software/pipeline/main_imageMatching.cpp [hunk 1]
```diff
@@ -92,51 +92,59 @@ std::ostream& operator<<(std::ostream& os, const OrderedPairList & pl)
...
+std::string EImageMatchingMode_description()
 {
-  return "The mode to combine matching between multiple SfM: \n"
-         "* a_ab : image matching for images in input SfMData A plus between A and B\n"
-         "* a_b  : image matching between input SfMData A and B\n";
+  return "The mode to combine matching: \n"
+         "* a+ab : image matching of images in SfMData A plus image matching between images of A and B\n"
+         "* aab  : image matching of images in SfMData A and between images of A and B\n"
+         "* ab   : image matching between images of SfMData A and B\n"
+         "* a    : image matching of images in SfMData A";
```
Comment (fabiencastan) [inline src/software/pipeline/main_imageMatching.cpp]:
```
return "The mode to combine image matching between the input SfMData A and B: \n"
           "* a|a+a|b : A with A + A with B\n"
           "* a|ab  : A with A and B\n"
           "* a|b   : A with B\n"
           "* a|a    : A with A";
```

Comment (gregoire-dl) [reply src/software/pipeline/main_imageMatching.cpp]: @fabiencastan I try it but, the pipe character gives a lot of trouble for the command line.

Comment (fabiencastan) [reply src/software/pipeline/main_imageMatching.cpp]: If it's inside quotes, it should work.
But you can use `/` instead.

File: src/software/pipeline/main_imageMatching.cpp [hunk 2]
```diff
@@ -391,7 +463,14 @@ int main(int argc, char** argv)
 
   // multiple SfM
   const bool useMultiSfM = !sfmDataFilenameB.empty();
-  const EImageMatchingMultiSfM modeMultiSfM = EImageMatchingMultiSfM_stringToEnum(modeMultiSfMName);
+  const EImageMatchingMode matchingMode = EImageMatchingMode_stringToEnum(matchingModeName);
+
+  if((useMultiSfM && (matchingMode == EImageMatchingMode::A_A)) ||
+     (!useMultiSfM && (matchingMode != EImageMatchingMode::A_A)))
```
Comment (fabiencastan) [inline src/software/pipeline/main_imageMatching.cpp]: `useMultiSfM == (matchingMode == EImageMatchingMode::A_A)`

File: src/software/pipeline/main_imageMatching.cpp [hunk 3]
```diff
@@ -459,135 +544,135 @@ int main(int argc, char** argv)
...
-    // create the database
-    ALICEVISION_LOG_INFO("Creating the database...");
+    // create the databases
+    ALICEVISION_LOG_INFO("Creating the databases...");
 
     // add each object (document) to the database
-    aliceVision::voctree::Database db(tree.words());
+    aliceVision::voctree::Database dbA(tree.words());
+    aliceVision::voctree::Database dbB;
+    aliceVision::voctree::Database dbAB;
```
Comment (fabiencastan) [inline src/software/pipeline/main_imageMatching.cpp]: Having 3 databases make things complicated.

You only need a 2nd database in *the* particular case A_A_AND_A_B. In all other cases, you just need to fill the DB according to the needs, and all the other code is the same.

Comment (gregoire-dl) [reply src/software/pipeline/main_imageMatching.cpp]: I personally prefer the 3 databases way. Found it more generic and comprehensible.
But I did it in a next commit.

File: src/software/pipeline/main_imageMatching.cpp [hunk 4]
```diff
@@ -652,14 +642,14 @@ int main(int argc, char** argv)
...
+          generateFromVoctree(allMatches, descriptorsFilesA, db2, tree, EImageMatchingMode::A_B, nbMaxDescriptors, numImageQuery);
         }
         break;
 
-        case EImageMatchingMode::A_AB: generateFromVoctree(allMatches, descriptorsFilesA, dbAB, tree, EImageMatchingMode::A_AB, nbMaxDescriptors, numImageQuery); break;
-        case EImageMatchingMode::A_B:  generateFromVoctree(allMatches, descriptorsFilesA, dbB,  tree, EImageMatchingMode::A_B,  nbMaxDescriptors, numImageQuery); break;
-        case EImageMatchingMode::A_A:  generateFromVoctree(allMatches, descriptorsFilesA, dbA,  tree, EImageMatchingMode::A_A,  nbMaxDescriptors, numImageQuery); break;
+        case EImageMatchingMode::A_AB: generateFromVoctree(allMatches, descriptorsFilesA, db, tree, EImageMatchingMode::A_AB, nbMaxDescriptors, numImageQuery); break;
+        case EImageMatchingMode::A_B:  generateFromVoctree(allMatches, descriptorsFilesA, db, tree, EImageMatchingMode::A_B,  nbMaxDescriptors, numImageQuery); break;
+        case EImageMatchingMode::A_A:  generateFromVoctree(allMatches, descriptorsFilesA, db, tree, EImageMatchingMode::A_A,  nbMaxDescriptors, numImageQuery); break;
```
Comment (fabiencastan) [inline src/software/pipeline/main_imageMatching.cpp]:
```
        case EImageMatchingMode::A_AB:
        case EImageMatchingMode::A_B:
        case EImageMatchingMode::A_A:
            generateFromVoctree(allMatches, descriptorsFilesA, db, tree, matchingMode,  nbMaxDescriptors, numImageQuery);
            break;
```

Comment (fabiencastan) [reply src/software/pipeline/main_imageMatching.cpp]: or just `if(matchingMode == EImageMatchingMode::A_A_AND_A_B) { xxx } else { xxx }`

File: src/software/pipeline/main_imageMatching.cpp [hunk 5]
```diff
@@ -594,18 +594,17 @@ int main(int argc, char** argv)
...
         }
 
-        if((matchingMode == EImageMatchingMode::A_A_AND_A_B) ||
+        if((matchingMode == EImageMatchingMode::A_AB) ||
            (matchingMode == EImageMatchingMode::A_B))
         {
-          nbFeaturesLoadedInputB = aliceVision::voctree::populateDatabase<DescriptorUChar>(sfmDataB, featuresFolders, tree, dbB, nbMaxDescriptors);
-          nbSetDescriptors +=  dbB.getSparseHistogramPerImage().size();
+          nbFeaturesLoadedInputB = aliceVision::voctree::populateDatabase<DescriptorUChar>(sfmDataB, featuresFolders, tree, db, nbMaxDescriptors);
+          nbSetDescriptors +=  db.getSparseHistogramPerImage().size();
```
Comment (fabiencastan) [inline src/software/pipeline/main_imageMatching.cpp]: As we accumulate in the same DB size() will return the full number of images and not only the ones from B, isn't it?

Comment (gregoire-dl) [reply src/software/pipeline/main_imageMatching.cpp]: yes, I just corrected in the next commit.
I hadn't seen it, `nbSetDescriptors` is only used for printing.

**Pull request merged**